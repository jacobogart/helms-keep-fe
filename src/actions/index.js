export const storeQuests = quests => ({
  type: 'STORE_QUESTS',
  payload: {
    quests
  }
})

export const addQuest = quest => ({
  type: 'ADD_QUEST',
  payload: {
    quest
  }
})

export const editQuest = quest => {
  return {
    type: "EDIT_QUEST",
    payload: {
      quest
    }
  };
}

export const deleteQuest = id => {
  return {
    type: "DELETE_QUEST",
    payload: {
      id
    }
  };
}

export const showPopup = (bool, id, type) => {
  return {
    type: "SHOW_POPUP",
    payload: {
      bool,
      id,
      type
    }
  };
}

export const addExp = (exp) => {
  return {
    type: 'ADD_EXP',
    payload: {
      exp
    }
  }
}
export const addLvl = (lvl) => {
  return {
    type: 'ADD_LVL',
    payload: {
      lvl
    }
  }
}
export const addNextLvl = (lvl) => {
  return {
    type: 'ADD_NEXT_LVL',
    payload: {
      lvl
    }
  }
}