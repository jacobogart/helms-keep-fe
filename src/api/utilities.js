export const host = 'http://localhost:5000/api/';

export const mockTask = {
  id: '1234',
  isCompleted: false, 
  message: 'Clean the kitchen'
};

export const mockCompletedTask = {
  id: '5678',
  isCompleted: true,
  message: 'Make a mock completed task'
}

export const mockQuest = {
	id: '1234',
  title: 'My First Quest',
  color: "green",
	challenges: [ mockTask, mockCompletedTask ]
};
