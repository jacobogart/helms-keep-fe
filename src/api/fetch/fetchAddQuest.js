import { host } from '../utilities';
import shortid from 'shortid';

export const fetchAddQuest = (title, challenges) => {
	const url = `${host}quests`;
	const body = { id: shortid.generate(), color: 'green', title, challenges };
	const options = {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(body)
	};
	return fetch(url, options).then(response => {
		if (!response.ok) {
			throw Error('Failed to add quest');
		} else {
			return response.json();
		}
	});
};
