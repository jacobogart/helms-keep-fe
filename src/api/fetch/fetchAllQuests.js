import { host } from '../utilities';

export const fetchAllQuests = () => {
	return fetch(`${host}quests`).then(response => {
		if (!response.ok) {
			throw Error('Failed to get quests');
		} else {
			return response.json();
		}
	});
};
