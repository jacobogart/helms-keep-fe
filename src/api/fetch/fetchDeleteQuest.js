import { host } from '../utilities';

export const fetchDeleteQuest = id => {
	const url = `${host}quests/${id}`;
	const init = {
		method: 'DELETE',
		headers: { 'Content-Type': 'application/json' }
	};
	return fetch(url, init).then(response => {
		if (!response.ok) {
			throw Error('Failed to delete quest');
		} else {
			return response;
		}
	});
};
