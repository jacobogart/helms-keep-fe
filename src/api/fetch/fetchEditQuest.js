import { host } from '../utilities';

export const fetchEditQuest = quest => {
	const url = `${host}quests/${quest.id}`;
	const options = {
		method: 'PUT',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(quest)
	};
	return fetch(url, options).then(response => {
		if (!response.ok) {
			throw Error('Failed to edit quest');
		} else {
			return response.json();
		}
	});
};
