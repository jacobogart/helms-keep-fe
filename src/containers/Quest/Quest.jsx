import React, { Component } from 'react';
import ChallengeContainer from '../../components/ChallengeContainer/ChallengeContainer';
import { fetchEditQuest } from '../../api/fetch/fetchEditQuest';
import { connect } from "react-redux";
import { editQuest, showPopup, addLvl, addExp } from '../../actions';
import { NavLink } from 'react-router-dom';
import PropTypes from 'prop-types';

export class Quest extends Component {
	constructor(props) {
		super(props);
		this.state = {
      showCompleted: false,
      showColors: false
		};
	}

	toggleShowCompleted = () => {
		this.setState({ showCompleted: !this.state.showCompleted });
  };

  addChallenge = challenge => {
    const quest = { ...this.props.data };
    quest.challenges.push(challenge);
    this.updateQuest(quest)
  };

  deleteChallenge = id => {
    const quest = { ...this.props.data };

    quest.challenges.splice(quest.challenges.findIndex(chal => chal.id === id), 1);
    this.updateQuest(quest);
  }

  updateChallenge = challenge => {
    const quest = { ...this.props.data };
    const chalIndex = quest.challenges.findIndex(chal => chal.id === challenge.id);
    quest.challenges[chalIndex] = challenge;
    this.updateQuest(quest);
  }
  
  showDeleteWarning = () => {
    const { id } = this.props.data;
    this.props.showPopup(true, id, 'delete');
  }

  updateQuest = quest => {
    this.props.updateQuest(quest);
    fetchEditQuest(quest);
  }

  updateTitle = e => {
    const quest = {...this.props.data};
    
    if (e.key === "Enter" || e.type === 'blur') {
      e.preventDefault();
      quest.title = e.target.innerText;
      this.updateQuest(quest);
      e.target.blur();
    }
  }
  
  pickColor = color => {
    const quest = { ...this.props.data };

    quest.color = color;
    this.updateQuest(quest);
    this.setState({ showColors: false });
  }

  render() {
    const { title, challenges } = this.props.data;
    const color = this.props.data.color || 'green';
    const completedChallenges = challenges.filter(chal => chal.isCompleted);
    const incompleteChallenges = challenges.filter(chal => !chal.isCompleted);
    const verb = !this.state.showCompleted ? '+ Show' : '- Hide';

    const completed = this.state.showCompleted && completedChallenges.length ?    <>
        <h4 className="divider">Completed</h4>
        <ChallengeContainer
          viewType="list"
          challengeType="complete"
          challenges={completedChallenges}
          addChallenge={this.addChallenge}
          updateChallenge={this.updateChallenge}
          deleteChallenge={this.deleteChallenge}
        />
      </>
      : null;

    const link = completedChallenges.length ?
      <p 
        role="button" 
        className="show-completed" 
        onClick={this.toggleShowCompleted}
      >
        {verb} {completedChallenges.length} completed
      </p>
      : null;
    
    const colorPicker = this.state.showColors ? (
      <div className="color-options">
        <button className="color-picker"><img className="coin" src={require(`../../assets/greencoin.png`)} type="button" onClick={() => this.pickColor('green')}/></button>
        <button className="color-picker"><img className="coin" src={require(`../../assets/yellowcoin.png`)} type="button" onClick={() => this.pickColor('yellow')}/></button>
        <button className="color-picker"><img className="coin" src={require(`../../assets/redcoin.png`)} type="button" onClick={() => this.pickColor('red')}/></button>
        <button className="color-picker"><img className="coin" src={require(`../../assets/bluecoin.png`)} type="button" onClick={() => this.pickColor('blue')}/></button>
        <button className="close-colors" type="button"onClick={() => this.setState({showColors: false})}><i className="fas fa-times"></i></button>
      </div>
    ) : null;
    
		return (
      <article className={`Quest ${color}-quest`}>
        <div className={`card-header ${color}-header`}>
          <button className="color-picker" type="button" onClick={() => this.setState({ showColors: !this.state.showColors })}><img className="coin title-coin" src={require(`../../assets/${color}coin.png`)}/></button>
          {colorPicker}
          <h2
            className="card-title"
            contentEditable="true"
            suppressContentEditableWarning={true}
            onKeyDown={this.updateTitle}
            onBlur={this.updateTitle}
          >
            {title}
          </h2>
          <NavLink to={`/quests/${this.props.data.id}`} key={this.props.data.id} className="expand-btn"><i className="fas fa-expand"></i></NavLink>
        </div>
        <div className="card-body">
          <ChallengeContainer
            viewType="list"
            challengeType="incomplete"
            challenges={incompleteChallenges}
            addChallenge={this.addChallenge}
            updateChallenge={this.updateChallenge}
            deleteChallenge={this.deleteChallenge}
          />
          {completed}
        </div>
        <div className={`card-footer ${color}-footer`}>
          <div>{link}</div>
          <button className="delete-btn" type="button">
            <i className="fas fa-trash" onClick={this.showDeleteWarning}/>
          </button>
        </div>
      </article>
    );
	}
}

Quest.propTypes = {
  data: PropTypes.object,
  showPopup: PropTypes.func,
  updateQuest: PropTypes.func
};
export const mapDispatchToProps = dispatch => ({
  updateQuest: quest => dispatch(editQuest(quest)),
  showPopup: (bool, id, type) => dispatch(showPopup(bool, id, type)),

})

export default connect(null, mapDispatchToProps)(Quest);
