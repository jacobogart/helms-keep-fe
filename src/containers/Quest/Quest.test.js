import React from 'react';
import { shallow } from 'enzyme';
import { Quest, mapDispatchToProps } from './Quest';
import { mockQuest, mockCompletedTask, mockTask } from '../../api/utilities';
import { fetchEditQuest } from '../../api/fetch/fetchEditQuest';
import { editQuest, showPopup } from '../../actions';

jest.mock('../../api/fetch/fetchEditQuest');

describe('Quest', () => {
	let wrapper, instance;

	const mockShowPopUp = jest.fn();
	const mockUpdateQuest = jest.fn();
	const mockKeyEvent = {
		key: 'Enter',
		preventDefault: () => {},
		target: {
			innerText: 'New title',
			blur: jest.fn()
		}
	};
	const mockBlurEvent = {
		type: 'blur',
		preventDefault: () => {},
		target: {
			innerText: 'New title',
			blur: jest.fn()
		}
	}

	beforeEach(() => {
		const mockQuest = {
			id: '1234',
			title: 'My First Quest',
			color: "green",
			challenges: [mockTask, mockCompletedTask]
		}
		wrapper = shallow(
		<Quest 
			data={mockQuest} 
			showPopup={mockShowPopUp}
			updateQuest={mockUpdateQuest}
		/>);
		instance = wrapper.instance();
	});

	it('should match snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});

	it('should have a default state', () => {
		const defaultState = {
			showCompleted: false,
			showColors: false
		};
		expect(wrapper.state()).toEqual(defaultState);
	});

	describe('Event Handlers', () => {
		it('should invoke toggleShowCompleted on click', () => {
			expect(wrapper.state('showCompleted')).toEqual(false);
			wrapper.find('.show-completed').simulate('click');
			expect(wrapper.state('showCompleted')).toEqual(true);
		});

		it('should invoke updateTitle when the title is edited', () => {
			jest.spyOn(instance, 'updateTitle');
			wrapper.find('.card-title').simulate('keydown', mockKeyEvent);
			expect(mockKeyEvent.target.blur).toHaveBeenCalled();
		});

		it('should invoke updateTitle when the title is blured', () => {
			jest.spyOn(instance, 'updateTitle');
			wrapper.find('.card-title').simulate('blur', mockBlurEvent);
			expect(mockKeyEvent.target.blur).toHaveBeenCalled();
		});

		it('should invoke showDeleteWarning when the delete button is clicked', () => {
			wrapper.find('.fa-trash').simulate('click');
			expect(mockShowPopUp).toHaveBeenCalledWith(true, '1234', 'delete')
		});
	});

	describe('toggleShowCompleted', () => {
		it('should toggle the showCompleted property in state', () => {
			expect(wrapper.state('showCompleted')).toEqual(false);
			instance.toggleShowCompleted();
			expect(wrapper.state('showCompleted')).toEqual(true);
			instance.toggleShowCompleted();
			expect(wrapper.state('showCompleted')).toEqual(false);
		});
	});

	describe('addChallenge', () => {
		it('should add a challenge to the challenges array of quest', () => {
			expect(instance.props.data.challenges).toHaveLength(2);
			instance.addChallenge(mockTask);
			expect(instance.props.data.challenges).toHaveLength(3);
		});

		it('should invoke updateQuest with the the new quest data', () => {
			jest.spyOn(instance, 'updateQuest');
			instance.addChallenge(mockTask);
			let localNote = { ...instance.props.data}
			localNote.challenges.push(mockTask);
			expect(instance.updateQuest).toHaveBeenCalledWith(localNote);
		});
	});

	describe('deleteChallenge', () => {
		it('should remove the appropriate challenge from the quest data', () => {
			expect(instance.props.data.challenges).toHaveLength(2);
			instance.deleteChallenge(5678);
			expect(instance.props.data.challenges).toHaveLength(1);
		});

		it('should invoke updateQuest with the the new quest data', () => {
			jest.spyOn(instance, 'updateQuest');
			instance.deleteChallenge(1234);
			let localNote = { ...instance.props.data }
			localNote.challenges.pop();
			expect(instance.updateQuest).toHaveBeenCalledWith(localNote);
		});
	});

	describe('updateChallenge', () => {
		let localChallenge = {
			id: '1234',
			isCompleted: false,
			message: 'update todo list'
		}
		it('should reassign the appropriate challenge', () => {
			instance.updateChallenge(localChallenge);
			expect(instance.props.data.challenges).toEqual([localChallenge, mockCompletedTask])
		});

		it('should invoke updateQuest with the update quest data', () => {
			jest.spyOn(instance, 'updateQuest');
			let localNote = { ...instance.props.data };
			localNote.challenges[0] = localChallenge;
			instance.updateChallenge(localChallenge);
			expect(instance.updateQuest).toHaveBeenCalledWith(localNote);
		});
	});

	describe('showDeleteWarning', () => {
		it('should invoke showPopup with the correct params', () => {
			instance.showDeleteWarning();
			expect(mockShowPopUp).toHaveBeenCalledWith(true, '1234', 'delete');
		});
	});

	describe('updateQuest', () => {
		it('should invoke the updateQuest dispatch with the correct params', () => {
			instance.updateQuest(mockQuest);
			expect(mockUpdateQuest).toHaveBeenCalledWith(mockQuest);
		});

		it('should invoke the fetchEditQuest dispatch with the correct params', () => {
			instance.updateQuest(mockQuest);
			expect(fetchEditQuest).toHaveBeenCalledWith(mockQuest);
		});
	});

	describe('updateTitle', () => {
		let updatedNote = { ...mockQuest, title: 'New title'};

		it('should invoke updateQuest if the user presses enter', () => {
			jest.spyOn(instance, 'updateQuest');
			instance.updateTitle(mockKeyEvent);
			expect(instance.updateQuest).toHaveBeenCalledWith(updatedNote);
		});

		it('should invoke updateQuest if the title blurs', () => {
			jest.spyOn(instance, 'updateQuest');
			instance.updateTitle(mockBlurEvent);
			expect(instance.updateQuest).toHaveBeenCalledWith(updatedNote);
		});

		it('should not invoke updateQuest if the condition is not met', () => {
			jest.spyOn(instance, 'updateQuest');
			instance.updateTitle({});
			expect(instance.updateQuest).not.toHaveBeenCalledWith();
		});
	});

	describe('MDTP', () => {
		let mappedProps, mockDispatch;
		beforeEach(() => {
			mockDispatch = jest.fn();
			mappedProps = mapDispatchToProps(mockDispatch);
		});

		it('should call dispatch when updateQuest is called', () => {
			const actionToDispatch = editQuest(mockQuest);
			mappedProps.updateQuest(mockQuest);
			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
		});

		it('should call dispatch when showPopup is called', () => {
			const actionToDispatch = showPopup(true, 1, 'delete');
			mappedProps.showPopup(true, 1, 'delete');
			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
		});
	});
});
