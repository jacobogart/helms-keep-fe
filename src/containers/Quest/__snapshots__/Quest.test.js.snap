// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quest should match snapshot 1`] = `
<article
  className="Quest green-quest"
>
  <div
    className="card-header green-header"
  >
    <button
      className="color-picker"
      onClick={[Function]}
      type="button"
    >
      <img
        className="coin title-coin"
        src="greencoin.png"
      />
    </button>
    <h2
      className="card-title"
      contentEditable="true"
      onBlur={[Function]}
      onKeyDown={[Function]}
      suppressContentEditableWarning={true}
    >
      My First Quest
    </h2>
    <NavLink
      className="expand-btn"
      key="1234"
      to="/quests/1234"
    >
      <i
        className="fas fa-expand"
      />
    </NavLink>
  </div>
  <div
    className="card-body"
  >
    <ChallengeContainer
      addChallenge={[Function]}
      challengeType="incomplete"
      challenges={
        Array [
          Object {
            "id": "1234",
            "isCompleted": false,
            "message": "Clean the kitchen",
          },
        ]
      }
      deleteChallenge={[Function]}
      updateChallenge={[Function]}
      viewType="list"
    />
  </div>
  <div
    className="card-footer green-footer"
  >
    <div>
      <p
        className="show-completed"
        onClick={[Function]}
        role="button"
      >
        + Show
         
        1
         completed
      </p>
    </div>
    <button
      className="delete-btn"
      type="button"
    >
      <i
        className="fas fa-trash"
        onClick={[Function]}
      />
    </button>
  </div>
</article>
`;
